### DISCUSSION TITLE
What causes hyper and hyper-based frameworks to have lower response times compared to actix-web?

### DISCUSSION THREAD

**Discussion by ishtms** [hyperium/hyper#2928] [General]
I saw the results on TFB and I was kinda surprised by how low the plaintext responses were on Hyper based frameworks. 

Also, to do some testing, I created a benchmark repo to test all the existing rust web frameworks on a simple "hello world" response text.

You can find the repository with all the results here - [Github Link](https://github.com/ishtms/rust-framework-benchmarks).

The tests are performed with rewrk and the most surprising element of the results was when the concurrency was set to "2048", and all the hyper based frameworks were 2x slower than actix-web.

Any reasons/explanations on why would this be an issue?

**Comment by seanmonstar** [comment]
I do not have an explanation, but I'd welcome if you were to investigate :)

**Comment by denosaurtrain** [comment]
@ishtms wow! Thanks for putting together these benchmark results! The community has really needed something like this. I note that Hyper has higher req/sec than Actix Web with some of the lower concurrency levels. Furthermore, the results with `2048` concurrency are worse for all frameworks compared to the same frameworks' `1024` concurrency tests, and the same pattern appears to be true for several of the other tests as well. At higher concurrency levels, the tests likely need more than `2` threads.

**Comment by ishtms** [comment]
@denosaurtrain Thanks for your response. I am testing this on M1 Max mac and this chip really has issues with wrk, rewrk and other benchmarking utilities. If I try to increase threads, the req/s falls down drastically. And if I try to use more than 10 threads, the benchmark doesn't even runs. Some weird issue.

On the other hand, these tests run perfectly fine if we set the thread count to default (`2`). 

If you notice, the `req/s` numbers start to go down starting from `128 concurrent connections` and upwards. The code for all the frameworks is inside `src/bin`. Is there anything you'd like me to change with how request is being handled on hyper?

Here's the hyper code that's being used

```rs
use std::convert::Infallible;
use hyper::service::{make_service_fn, service_fn};
use hyper::{Body, Request, Response, Server};

async fn hello(_: Request<Body>) -> Result<Response<Body>, Infallible> {
    Ok(Response::new("Hello world!".into()))
}

#[tokio::main]
pub async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let make_svc = make_service_fn(|_conn| async { Ok::<_, Infallible>(service_fn(hello)) });
    let port_number: u16 = str::parse(get_port_number().as_str()).unwrap();

    let addr = ([127, 0, 0, 1], port_number).into();
    let server = Server::bind(&addr).serve(make_svc);
    server.await?;

    Ok(())
}

fn get_port_number() -> String {
    std::env::args().collect::<Vec<String>>()[1].clone()
}
```

**Reply by denosaurtrain** [reply]
It seems you may be controlling the number of threads for the client but not the server. For frameworks that sit on top of Tokio, this might help: https://docs.rs/tokio/1.20.1/tokio/attr.main.html#multi-threaded-runtime

**Reply by denosaurtrain** [reply]
Benchmarking on a laptop can be tricky. Issue 1: thermal throttling. Issue 2: the M1 appears to have some cores dedicated for "efficiency" and others for "performance."

**Reply by denosaurtrain** [reply]
According to a random source:
> The 10-core CPU has eight high-performance cores and two high-efficiency cores. 

If I were you, (and to be clear all of this is guesswork without tons of testing) I would keep 2 cores for the benchmarking client, then constrain each of the web frameworks to 4 cores, leaving plenty of overhead for other noise on your system. I'd also find some way to monitor thermals during the tests -- IDK what's out there for Macs, but there should be something simple I would hope.

**Reply by ishtms** [reply]
Seems like tokio is actually starting 10 threads even without specifying `(flavor="multi_thread",  worker_threads=10)`. I am also using "High performance"mode in the settings before running the benchmarks, and that makes sure the processes utilise high-performance cores over efficiency cores.

<img width="1197" alt="image" src="https://user-images.githubusercontent.com/30657558/181397611-0d32b3b4-76c8-4fff-b1b5-6032646cda74.png">

**Reply by ishtms** [reply]
I am going to do another benchmark and update you when the results are up, to see if there's any changes - 60 secs each for 16, 32, 64, 128, 256, 512 and 1024. 2048 doesn't makes sense anyways on m1 chip so I'll drop that.

**Reply by ishtms** [reply]
@denosaurtrain The new results are up - [here](https://github.com/ishtms/rust-framework-benchmarks). The gap has reduced significantly now

**Reply by ishtms** [reply]
And, hyper is performing much better than actix-web in almost all benchmarks.
